### PR TITLE
Fixed NaN's GetHashCode and Equals invariant issue.

### LINKED ISSUE #6237: double.GetHashCode and double.Equals do not agree (When NaN and -NaN are compared for equality)
It is an invariant that two objects for which `o1.Equals(o2)`, their hash codes should agree, i.e. `o1.GetHashCode() == o2.GetHashCode()`. See [the specification](https://msdn.microsoft.com/en-us/library/system.object.gethashcode%28v=vs.110%29.aspx), which says

> Two objects that are equal return hash codes that are equal.

This invariant is assumed by many APIs that use these methods, and any types that fail to satisfy them are "broken" and cannot be expected to work with the core APIs. However, the implementation for `double` does not satisfy this invariant. Specifically, the following snippet of code prints "True False".

``` cs
        object o1 = double.NaN;
        object o2 = -double.NaN;
        Console.WriteLine(o1.Equals(o2));
        Console.WriteLine(o1.GetHashCode() == o2.GetHashCode());
```

### PR DESCRIPTION
**Pull request by mazong1123**
This PR keeps the invariant of GetHashCode and Equals for NaNs. We always return 1 for GetHashCode of NaNs so that all NaNs have the same hash code.

Fix #6237

### CODE REVIEW

File: src/mscorlib/shared/System/Single.cs [hunk 1]
```diff
@@ -230,6 +230,13 @@ public unsafe override int GetHashCode()
                 // Ensure that 0 and -0 have the same hash code
                 return 0;
             }
+
+            if (IsNaN(f))
+            {
+                // Ensure that all NaNs have the same hash code
+                return 1;
```
Comment (tannergooding) [inline src/mscorlib/shared/System/Single.cs]: This should just be `f = float.NaN;`, then the subsequent code will do the right thing. We don't want to use 1, as that will cause definite conflicts between NaN and `1.40129846E-45`.

File: src/mscorlib/shared/System/Double.cs [hunk 2]
```diff
@@ -238,6 +238,13 @@ public unsafe override int GetHashCode()
                 // Ensure that 0 and -0 have the same hash code
                 return 0;
             }
+
+            if (IsNaN(d))
+            {
+                // Ensure that all NaNs have the same hash code
+                return 1;
```
Comment (tannergooding) [inline src/mscorlib/shared/System/Double.cs]: This should just be `d = double.NaN;`, then the subsequent code will do the right thing. We don't want to use 1, as that will cause definite conflicts between NaN and `4.9406564584124654E-324`.

**Comment by mazong1123** [comment]
@tannergooding I've updated accordingly. Thanks for your review.

**Comment by tannergooding** [comment]
Thanks! Someone on the actual CoreCLR team will need to review/approve/merge.

**Comment by mazong1123** [comment]
@dotnet-bot test Tizen armel Cross Debug Build

**Comment by jkotas** [comment]
How much slower is Double.GetHashCode with this change? Have you tried different variants and picked the fastest one to minimize the perf regression from this fix?

**Comment by danmosemsft** [comment]
@mazong1123 thanks, before merging anything in Corelib, we like to have a PR up in CoreFX with test(s) that we can merge when the changes gets there.

**Comment by mazong1123** [comment]
@jkotas 
> How much slower is Double.GetHashCode with this change?

Will run a micro-benchmark today.

>  Have you tried different variants and picked the fastest one to minimize the perf regression from this fix?

For perf regression, are you saying the hash conflict? I thought use `float.NaN/double.NaN` to calculate the hash code for all NaNs is the best choice to minimize conflict.

**Comment by mazong1123** [comment]
@danmosemsft The PR with tests is already there: https://github.com/dotnet/corefx/pull/23793

**Comment by jkotas** [comment]
I meant how much Double.GetHashCode will get slower when there are no NaNs involved at all.

**Comment by tannergooding** [comment]
@jkotas, it's probably barely measurable (we should get numbers to confirm though). IsNaN is a single bitwise comparison and jump.

It would actually be branchless if we supported `CMOVcc` instructions, but I don't believe we do.

**Comment by mazong1123** [comment]
Following are the micro-benchmark results:

**Before Change**
Start collecting double.GetHashCode() performance...
280
278
279
279
278
279
279
279
278
278
279
279
279
279
279
279
279
279
278
278
Start collecting float.GetHashCode() performance...
430
429
429
430
429
429
428
429
429
429
430
429
429
429
429
430
429
429
430
429
Done.

**After Change**
Start collecting double.GetHashCode() performance...
349
348
348
348
355
348
347
347
348
348
347
347
347
348
347
348
347
347
347
347
Start collecting float.GetHashCode() performance...
331
331
332
331
331
332
332
332
371
464
331
332
332
331
331
331
332
333
332
332
Done.

** Summary **
`double.GetHashCode()` 's performance has 20% regression. However, the most interested thing is `float.GetHashCode()` 's performance has 23% improvement - I ran this tests several times and get the same result. Following is my micro-benchmark code: 

```csharp
using System;
using System.Diagnostics;

namespace hashequalagree
{
    class Program
    {
        private static volatile bool _doubleHashCodeEquality = false;
        private static volatile bool _floatHashCodeEquality = false;

        static void Main(string[] args)
        {
			Console.WriteLine("Start collecting double.GetHashCode() performance...");

            double o1 = 12345;
            double o2 = -24687;

            for (int i = 0; i < 20; i++)
            {
                Stopwatch watch = new Stopwatch();
                watch.Start();
                for (int j = 0; j < 10000000; j++)
                {
                    _doubleHashCodeEquality = o1.GetHashCode() == o2.GetHashCode();
					_doubleHashCodeEquality = o1.GetHashCode() == o2.GetHashCode();
					_doubleHashCodeEquality = o1.GetHashCode() == o2.GetHashCode();
					_doubleHashCodeEquality = o1.GetHashCode() == o2.GetHashCode();
                }

                Console.Write(watch.ElapsedMilliseconds);
                Console.WriteLine();
            }

			Console.WriteLine("Start collecting float.GetHashCode() performance...");

            float f1 = 123f;
            float f2 = -256f;

            for (int i = 0; i < 20; i++)
            {
                Stopwatch watch = new Stopwatch();
                watch.Start();
                for (int j = 0; j < 10000000; j++)
                {
                    _floatHashCodeEquality = f1.GetHashCode() == f2.GetHashCode();
					_floatHashCodeEquality = f1.GetHashCode() == f2.GetHashCode();
					_floatHashCodeEquality = f1.GetHashCode() == f2.GetHashCode();
					_floatHashCodeEquality = f1.GetHashCode() == f2.GetHashCode();
                }

                Console.Write(watch.ElapsedMilliseconds);
                Console.WriteLine();
            }

			Console.WriteLine(_doubleHashCodeEquality);
			Console.WriteLine(_floatHashCodeEquality);

			Console.WriteLine("Done.");
        }
    }
}

```

**Comment by mazong1123** [comment]
I found the problem. I'm comparing to stable .net core 2.0.0 release whereas my changes are based on master branch code. I need to compare to the codebase which I apply the changes to. So the new micro benchmark results are:

**Before Change**

Start collecting double.GetHashCode() performance...
302
303
305
312
323
324
325
341
322
350
357
305
320
323
321
305
304
303
337
334
Start collecting float.GetHashCode() performance...
278
304
300
285
291
301
296
286
278
279
278
278
278
278
278
284
279
278
278
278
False
False
Done.

**After Change**

Start collecting double.GetHashCode() performance...
349
348
348
348
349
349
348
349
348
349
352
354
348
348
348
350
349
348
348
349
Start collecting float.GetHashCode() performance...
365
366
364
365
364
364
364
365
365
364
364
364
366
364
365
364
364
364
365
365
False
False
Done.

**Comment by mazong1123** [comment]
@jkotas @tannergooding Seems like there's no way to optimize it in managed code at least. Probably cannot be optimized in C++ code either. We probably need some asm hack to boost the performance but I'm wodering is it worth to investigate to it. Thoughts?

**Comment by jkotas** [comment]
I think there are ways to optimize this in managed code. E.g. I would expect that `if (IsNaN(d)) return 2439875943895 (put the actual NaN hash code here);` would be faster that what you are doing now. It avoids data flow dependency that is making the code slower than necessary.

**Comment by mikedn** [comment]
The code generated for something like `if (isnan) d = nan` is
```asm
       C4E17B110424         vmovsd   qword ptr [rsp], xmm0
       488B0424             mov      rax, qword ptr [rsp]
       48BAFFFFFFFFFFFFFF7F mov      rdx, 0x7FFFFFFFFFFFFFFF
       4823C2               and      rax, rdx
       48BA000000000000F07F mov      rdx, 0x7FF0000000000000
       483BC2               cmp      rax, rdx
       7E09                 jle      SHORT G_M55888_IG03
       C4E17B100515000000   vmovsd   xmm0, qword ptr [reloc @RWD00]
```
Unfortunately it involves storing/loading to/from memory and that's probably impossible to avoid without JIT support. The "normal" way to do this in ASM is `movq rax, xmm0`.

**Comment by mikedn** [comment]
Also, the nan check itself seems dubious. We don't care about the mantissa bits so it should be `x & 0x7FF0000000000000`, not `x & 0x7FFFFFFFFFFFFFFF`. So a shorter implementation might be
```asm
shl rax, 1
shr rax, 52
cmp eax, 0x7FFF
jne IG03
```

**Comment by mazong1123** [comment]
> I would expect that if (IsNaN(d)) return 2439875943895 (put the actual NaN hash code here); would be faster that what you are doing now

@jkotas `IsNaN()` is the bottleneck and it still will be executed in your example so the performance may not be increased much IMO (I'll verify it later).

**Comment by mazong1123** [comment]
I'm going to do a POC in low-lever - move the `IsNaN` implementation to native code. And mix some ASM hacks according to @mikedn 's comment.

**Comment by mikedn** [comment]
> move the IsNaN implementation to native code

That's unlikely to work, the cost of calling the native code will likely be higher than the actual IsNan code. At best, you can experiment entirely with native code to see if alternative implementations are measurably better.

**Comment by mazong1123** [comment]
I'm still trying to reduce the performance regression. @jkotas what's the expectation of this bug fix? I think it's hard to eliminate the perf regression unless JIT can "cleverly" skip `IsNaN` comparison. Even if I add some asm hacks to reduce the impact, we cannot eliminate the perf regression because we cannot avoid the `IsNaN` comparison. It almost falls to JIT optimazation if we want to eliminate the perf regression.

**Comment by jkotas** [comment]
I understand that this fix will make things slower, everything else being equal. I do not think that writing this in assembly, etc. is a good idea. My ask was to try different variants of the C# implementation to minimize the regression. For example, I would expect that implementation like this:

```
if (IsNaN(d))
{
    // Ensure that all NaNs have the same hash code
    return 0xXXXXXXXXX;
}
```

would reduce the perf regression.

**Comment by tannergooding** [comment]
> Also, the nan check itself seems dubious. We don't care about the mantissa bits so it should be x & 0x7FF0000000000000, not x & 0x7FFFFFFFFFFFFFFF.

@mikedn, the code we use is the same code MSVC and GCC both use in their `isnan` check.

> So a shorter implementation might be
```
shl rax, 1
shr rax, 53
cmp eax, 0x7FFF
jne IG03
```

I would think we should treat it as an intrinsic and instead generate:
```
xor eax, eax
ucomisd xmm0, xmm0
setp al
```

**Comment by tannergooding** [comment]
Sorry, meant to post this as well.

I don't think doing the below will work, since that will miss out on 0x7FF0000000000001
through 0x7FFEFFFFFFFFFFFF, which are also valid NaN representations.
```
shl rax, 1
shr rax, 53
cmp eax, 0x7FFF
jne IG03
```

**Comment by mikedn** [comment]
> I don't think doing the below will work, since that will miss out on 0x7FF0000000000001
through 0x7FFEFFFFFFFFFFFF, which are also valid NaN representations.

It looks to me that it's the other way around, it doesn't work because it includes INFs, not because it misses NaNs. Though perhaps including INFs wouldn't be a problem in this particular context. Come to think of it, I'm not sure if this problem makes sense to begin with. Putting a NaN in a hashtable is nonsense.

> I would think we should treat it as an intrinsic and instead generate:

Until an intrinsic a simple equality check (x != x) could do. The generated code is slightly larger but it's better than the original code. But this approach may run into trouble with SNaNs if someone unmasks FP exceptions.

**Comment by tannergooding** [comment]
> Come to think of it, I'm not sure if this problem makes sense to begin with. Putting a NaN in a hashtable is nonsense.

I agree, but we still have to follow the contract 😄.

> Until an intrinsic a simple equality check (x != x) could do.

That'll probably work here. We should add a bug tracking better codegen for `double.IsNaN` and `float.IsNaN` proper though.

> But this approach may run into trouble with SNaNs if someone unmasks FP exceptions.

For what it's worth, GCC, Clang, and MSVC generate the `ucomisd` and `ucomisf` instruction for `isnan` and `isnanf`, when intrinsics are enabled.

I think that someone changing the FP exception reporting behavior somewhat falls into the area of "undefined behavior", although I am not sure if there are any existing checks/assumptions/codgen in the runtime that expect exceptions to be masked or not.

**Comment by karelz** [comment]
What is status of this PR? Is anyone still working on it? What are the next steps?

cc @valenis

**Comment by mazong1123** [comment]
I'm working on it. I tried to figure out a way to mitigate the regression as @jkotas said but it seems like failed. @jkotas do we still need to try more varants or are we good with current perf?

**Comment by karelz** [comment]
Would it make sense to add a perf test(s) for the measurements above. Something that can be validated independently?

**Comment by mazong1123** [comment]
> Would it make sense to add a perf test(s) for the measurements above

Yes. But we should add the perf tests after this PR get merged IMO. 

I think we need to revert the last commit because it makes the perf worse. And I still think the only way to mitigate the regression is to improve jit.

**Comment by karelz** [comment]
If we have questions about perf, the test should be definitely part of the PR.
If you think JIT work is needed first, I would suggest to close this PR until we have the changes in JIT.

**Comment by tannergooding** [comment]
@mazong1123, did you end up checking perf when you check for NaN via `x != x`?

**Comment by mazong1123** [comment]
@tannergooding  are you saying change [the implementation of `IsNaN`](https://github.com/mazong1123/coreclr/blob/12a99be9352ee87ca410dee82512e4286250ede5/src/mscorlib/shared/System/Double.cs#L72-L73) to `x != x` ? Not yet, I'll test it and come back.

**Comment by jkotas** [comment]
@mazong1123 I am not able to reproduce your results. I see a large difference between baseline and your change because your change made the method to go over inlining threshold. And the difference between the last commit and last commit reverted is much smaller for me (the native code with last commit is smaller and looks better, but it runs slower for some reason).

  | Double | Float
-- | -- | --
Baseline | 103 | 79
HEAD | 236 | 132
HEAD-1 | 235 | 129

**Comment by mazong1123** [comment]
> And the difference between the last commit and last commit reverted is much smaller for me (the native code with last commit is smaller and looks better, but it runs slower for some reason).

It sounds like the same result as mine - last commit looks better but actually slower than the HEAD - 1 one, but the difference is small compare to the difference to the baseline.

I'm trying to change `IsNaN` implementation to `x != x` as @tannergooding . It looks promising although it introduces a warning. I'll come back after more measurements done.

**Comment by tannergooding** [comment]
@mazong1123, sorry if I wasn't clear. I meant just update the check locally (the code in `GetHashCode`) rather than the actual `Double.IsNaN` implementation.

The actual `IsNaN` method should probably remain as-is (since that is the same code used by other implementations, MSVC, GCC, etc) and the runtime should be updated to treat it as intrinsic and generate `ucomisd` . Updating it to be an intrinsic is a separate work item, however (https://github.com/dotnet/coreclr/issues/14846).

**Comment by mazong1123** [comment]
OK I got the average measurement on my machine. The `x != x` comparison is better than the leveraging `IsNaN`:

  | Double | Float
-- | -- | --
Baseline | 305 | 279
HEAD - 1 | 382 | 340
x != x | 343 | 334

We still have regression, so we may expect #14846 to improve the perf further. Currently, I won't use `IsNaN` inside `GetHashCode` and instead, use `x != x` for NaN check. Please let me know do we need to wait for #14846 or just use `x != x` to get the NaN hash code issue fixed. @jkotas @tannergooding

**Comment by mazong1123** [comment]
@tannergooding @jkotas  Do we need to wait for #14846 ? If not, do you have any time to review this change? Thanks.

**Comment by jkotas** [comment]
We need to understand performance impact of the change, and be comfortable with taking the regression.

When I run the microbenchmark from https://github.com/dotnet/coreclr/pull/13781#issuecomment-327978641 a few days ago, I saw 2.3x regression - https://github.com/dotnet/coreclr/pull/13781#issuecomment-341696845. It is pretty different from the numbers in https://github.com/dotnet/coreclr/pull/13781#issuecomment-341867230.

@tannergooding What do the numbers look like for you?

**Comment by mazong1123** [comment]
@jkotas I updated the code to leverage `x != x` pattern to determine `NaN` last Friday. Do you have time to run the micro-benchmark using current code?

**Comment by tannergooding** [comment]
@jkotas:
### master

```
Start collecting double.GetHashCode() performance (100 iterations)...
    Min Elapsed Time: 86 ms
    Avg Elapsed Time: 91 ms
    Max Elapsed Time: 138 ms
    Ttl Elapsed Time: 9141 ms

Start collecting float.GetHashCode() performance (100 iterations)...
    Min Elapsed Time: 78 ms
    Avg Elapsed Time: 82 ms
    Max Elapsed Time: 113 ms
    Ttl Elapsed Time: 8288 ms
```

### x != x

```
Start collecting double.GetHashCode() performance (100 iterations)...
    Min Elapsed Time: 169 ms
    Avg Elapsed Time: 179 ms
    Max Elapsed Time: 246 ms
    Ttl Elapsed Time: 17978 ms

Start collecting float.GetHashCode() performance (100 iterations)...
    Min Elapsed Time: 125 ms
    Avg Elapsed Time: 131 ms
    Max Elapsed Time: 163 ms
    Ttl Elapsed Time: 13121 ms
```

### IsNaN

```
Start collecting double.GetHashCode() performance (100 iterations)...
    Min Elapsed Time: 176 ms
    Avg Elapsed Time: 183 ms
    Max Elapsed Time: 229 ms
    Ttl Elapsed Time: 18363 ms

Start collecting float.GetHashCode() performance (100 iterations)...
    Min Elapsed Time: 134 ms
    Avg Elapsed Time: 139 ms
    Max Elapsed Time: 178 ms
    Ttl Elapsed Time: 13950 ms
```

### x != x, return precomputed hash

```
Start collecting double.GetHashCode() performance (100 iterations)...
    Min Elapsed Time: 133 ms
    Avg Elapsed Time: 138 ms
    Max Elapsed Time: 178 ms
    Ttl Elapsed Time: 13827 ms

Start collecting float.GetHashCode() performance (100 iterations)...
    Min Elapsed Time: 125 ms
    Avg Elapsed Time: 131 ms
    Max Elapsed Time: 187 ms
    Ttl Elapsed Time: 13197 ms
```

@mazong1123, I updated the microbench a bit to print some other useful info and be more easily modifiable: [Program.cs.txt](https://github.com/dotnet/coreclr/files/1450645/Program.cs.txt)

**Comment by tannergooding** [comment]
@mazong1123, seems like `x != x, with precomputed NaN hash` is the best option, if the change is taken.

However, it is still a 52% (double) and 60% (single) regression. So it may be that we need to wait for better codegen.

NOTE, by precomputed hash, I mean doing:
```
if (d != d)
{
    // Ensure that all NaNs have the same hash code
    return -524288;
}
```

and

```
if (f != f)
{
    // Ensure that all NaNs have the same hash code
    return -4194304;
}
```

**Comment by mazong1123** [comment]
@karelz Based on the measurement and the discussions with @jkotas and @tannergooding , we'd better wait for #14846 . This PR may be hung for a long time so I'm fine if you want to close this PR for now :)

**Comment by mikedn** [comment]
I'll try to do a quick and dirty implementation for 14846 to see if it can actually improve perf. I'm a bit skeptical about that, we're talking about removing a single branch that, at least in your tests, it always taken or always not taken.

**Comment by tannergooding** [comment]
@mikedn, you're probably right about that. I was just comparing the MSVC codegen to the current CLR codegen and they are pretty similar...

Notable differences:
* Our initialization code for loading scalar values into XMM registers isn't that great
  * We are writing zero to [rsp], then overwriting it with [rcx] (via xmm0) two instructions later
  * As part of that, we are writing xmm0 to [rsp] then immediately reloading [rsp] into xmm0
* When doing the `d != d` comparison, we are reloading `d` into xmm0/xmm1 when the value is already in xmm1
  * We don't even need two registers here, comparing `xmm1` to itself is fine
  * Even if using two registers was required, we are doing this by two memory reads from the same location, rather than a read and a register copy
* On the `d != d` comparison, we are branching to different locations for parity and equality. We could be jumping to the same location for parity and non-equality, which should help the branch predictor (IIRC)

MSVC:
```asm
entry:
  vmovsd      xmm1, QWORD PTR [rcx+8]
  vxorpd      xmm0, xmm0, xmm0
  vucomisd    xmm1, xmm0
  vmovsd      QWORD PTR [rsp], xmm1
  jp          label1
  jne         label1
  xor         eax, eax
  ret         0

label1:
  vucomisd    xmm1, xmm1
  jp          SHORT label2
  jne         SHORT label3
  mov         rcx, QWORD PTR [rsp]
  mov         rax, rcx
  sar         rax, 20h
  xor         eax, ecx
  ret         0
  
label2:
  mov         eax, 0FFF80000h
  ret         0
```

CoreCLR
```asm
entry:
  push        rax
  vzeroupper
  xor         eax, eax
  mov         QWORD PTR [rsp], rax
  vmovsd      xmm0, QWORD PTR [rcx]
  vmovsd      QWORD PTR [rsp], xmm0
  vmovsd      xmm0, QWORD PTR [rsp]
  vxorpd      xmm1, xmm1, xmm1
  vucomisd    xmm0, xmm1
  jp          label1
  jne         label1
  xor         eax, eax
  add         rsp, 8
  ret


label1:
  vmovsd      xmm0, QWORD PTR [rsp]
  vmovsd      xmm1, QWORD PTR [rsp]
  vucomisd    xmm0, xmm1
  jp          label2
  je          label3

label2:
  mov         eax, 0FFF80000h
  add         rsp, 8
  ret

label3:
  mov         rax, QWORD PTR [rsp]
  mov         rdx, rax
  sar         rdx, 20h
  xor         eax, edx
  add         rsp, 8
  ret
```

**Comment by mikedn** [comment]
> We are writing zero to [rsp], then overwriting it with [rcx] (via xmm0) two instructions later

I need to play a bit with it, the generated code looks fishy. It may be affected by the reinterpretation of float as int by taking the address of the float variable. Improving that may be more useful than improving `x != x` but it is also more complicated.

**Comment by mikedn** [comment]
Yeah, this mess is due to the address of the lcl var being taken. I'm not quite sure what's the best solution. If GetHashCode is not inlined then the following should work well:
```C#
if (m_value == 0)
    return 0;
if (m_value != m_value)
     return unchecked((int)0xFFF80000);
long value = Unsafe.As<double, long>(ref m_value);
return unchecked((int)value) ^ ((int)(value >> 32));
```
If `GetHashCode` is inlined then this still seems to generate better code than the original version but to be on the safe side you should still use a local variable:
```C#
if (m_value == 0)
    return 0;
if (m_value != m_value)
    return unchecked((int)0xFFF80000);
double d = m_value;
long value = Unsafe.As<double, long>(ref d);
return unchecked((int)value) ^ ((int)(value >> 32));
```
In general it's best to avoid using a local variable that has had its address taken for other things.

Note that the first version requires `Unsafe.As`, I haven't checked if that particular overload is available in corelib. The second version can continue to use a pointer, it shouldn't make any difference.

**Comment by mazong1123** [comment]
@jkotas @tannergooding  Do we need to wait for the JIT improvement for this change? If so should we close this PR right now?

**Comment by mikedn** [comment]
I doubt that any JIT improvements around `x != x` will make any difference. I'll try to run some benchmarks later today, I made the JIT changes a while ago.

**Comment by mikedn** [comment]
As far as I can tell, the impact of the extra branch generated by `x != x` is ~2%. And that's when you measure `GetHashCode` alone, if you measure something like `Dictionary.ContainsKey` then it's even smaller. So improving `x != x` in the JIT, while nice to have, won't make much difference in this case.

**Comment by mikedn** [comment]
I've patched the JIT to not go trough memory when using code like `long value = Unsafe.As<double, long>(ref d);` (by using `movd`). However, `Dictionary.ContainsKey` appears to be getting slower if I do that. That's odd, I'll need to check what's going on but I have other JIT stuff that I want to finish first so I won't be looking at this again for the next 1-2 weeks.

Asm code that we'd get with `movd`:
```asm
```asm
G_M33193_IG01:
       C5F877               vzeroupper
G_M33193_IG02:
       C4E17B1001           vmovsd   xmm0, qword ptr [rcx]
       C4E17057C9           vxorps   xmm1, xmm1
       C4E1792EC1           vucomisd xmm0, xmm1
       7A05                 jpe      SHORT G_M33193_IG04
       7503                 jne      SHORT G_M33193_IG04
       33C0                 xor      eax, eax
G_M33193_IG03:
       C3                   ret
G_M33193_IG04:
       C4E1792EC0           vucomisd xmm0, xmm0
       7A02                 jpe      SHORT G_M33193_IG05
       7409                 je       SHORT G_M33193_IG06
G_M33193_IG05:
       C4E17B100515000000   vmovsd   xmm0, qword ptr [reloc @RWD00]
G_M33193_IG06:
       C4E1F97EC0           vmovd    rax, xmm0
       488BD0               mov      rdx, rax
       48C1EA20             shr      rdx, 32
       33C2                 xor      eax, edx
G_M33193_IG07:
       C3                   ret
```

**Comment by jkotas** [comment]
Superseded by #16551

**Pull request closed without merging**